### ISSUE TITLE
Generic param in function looses `Literal` type

### ISSUE THREAD

**Issue by rayansostenes** [microsoft/pyright#4872]
**Describe the bug**
I'm having some problems with the following code snippet on pyright, it appears that when the type narrowing is lost when
we use the second param with the same generic.

I tried using `cast` but the results are the same

> PS: Sorry for the bad title, I really don't know how to explain what is happening

```python
from typing import Iterable, Literal, Sequence, TypeVar

T = TypeVar("T")
U = TypeVar("U")


def new_zip_longest(
    seq1: Sequence[T],
    seq2: Sequence[U],
    fill_value_1: T,
    fill_value_2: U,
) -> Iterable[tuple[T, U]]:
    ...


MyLiteralString = Literal["a", "b", "c", "d", "e", "f"]
MyLiteralInt = Literal[1, 2, 3, 4, 5, 6]


def func() -> Sequence[MyLiteralString]:
    lst_str: Sequence[Literal["a", "b", "c"]] = ["a", "b", "c"]
    lst_int: Sequence[MyLiteralInt] = [1, 2, 3]

    result = [
        string
        for string, _number in new_zip_longest(
            lst_str,
            lst_int,
            fill_value_1="a",
            fill_value_2=1,
        )
    ]
    reveal_type(result)
    return result

```

**Expected behavior**
Type of `result` should be `list[Literal['a', 'b', 'c']]`, that is compatible with the return type

[Pyright Playground](https://pyright-playground.decorator-factory.su/?gzip=H4sIAI7-JWQC_61TPU_DMBDd_StOYUkkU6nlY6gIEgMDEkxNWKooSttLseTawXZA5ddjOx91aWEiw0l-53f3zvdSK7kDZlAZKbkGtmukMvDFmpJLsUVtSO1umH3DxHZIP9n71YojhWdHrTiFBb63KNYWyvYNvlaKkAzS4RBHWZSQPARyCxCywRoEfpZBw5iA_TS-T-dj1WVW0AGeBXDewzXjvPyoeIulZWUnoOXklCRweT9qX5q2sTGjkBfF3BNYDRxF7FonkKbDYZZ0afcpNK0S7n38NeoFJT6N_Ih__zd9nPZQhnrFXnAaak8ALtwCcA5sK6TCc-3u_qvd9Ld2GoPCFdMID1pb3zApHpWSdqWLN9nyDQhpYIW2ebV-w41f88u-N8rCKOejdDDOMqoiCtHKhbULGxfQhToqDrwnYQKSncDKv6JwTeGGwm3RO6luxTr2Wx4d8qNzv2muTamNmkNngrNiCgqTyaSwbeMjPBkrMGGGCqHQA6_XaV-g24ZuuRtjOb6j9qLGYy1VD1EoRbtboQImzv8gw9ePQk9Aq-4YDJecupF-y87S6SHXjVv0E3xgxUtnjribJiGBzzqIfAMtp5ZQUwQAAA%3D%3D)

[MyPy Playground](https://mypy-play.net/?mypy=latest&python=3.11&gist=ed1ce08a813945abce2733efb472e187)

**Comment by erictraut** [comment]
Yes, this behavior is intended. Pyright's constraint solver does not retain literal types unless a literal is required to meet the constraints. Mypy is inconsistent in how it treats literals in its constraint solver. See [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=constraint-solver-literals) for details.

You can influence the result by using bidirectional type inference:
```python
    val: Iterable[tuple[MyLiteralString, MyLiteralInt]] = new_zip_longest(
        lst_str,
        lst_int,
        fill_value_1="a",
        fill_value_2=1,
    )

    result: list[MyLiteralString] = [string for string, _ in val]
```

**Comment by rayansostenes** [comment]
Thanks @erictraut. 

I did not know that the documentation had that page, it will help me a lot.
